infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java

```diff
@@ -96,6 +97,26 @@ private void applyDevfileVolumes(InternalMachineConfig machineConfig) {
     }
   }
 
+  private void applyDevfileEndpoints(InternalMachineConfig machineConfig)
+      throws InfrastructureException {
+    for (org.eclipse.che.api.core.model.workspace.devfile.Endpoint endpoint :
+        component.getEndpoints()) {
+      if (!machineConfig.getServers().containsKey(endpoint.getName())) {
+        machineConfig
+            .getServers()
+            .put(endpoint.getName(), ServerConfigImpl.createFromEndpoint(endpoint));
+      } else {
+        throw new InfrastructureException(
+            format(
+                "Devfile's component '%s' is trying to override plugin/editor's endpoint '%s'. "
```

Comment:
Nitipicking, but the following sounds a little bit cleaner to me (watch out for the swapping of the params):
`Devfile overrides the endpoint '%s' of the plugin/editor component '%s'. This is not allowed because it would most probably cause the workspace malfunction. Please change the name of the endpoint in the devfile and try to start the workspace again.`

Reply:
yes, better. updated. Thanks!